tests/phpunit/includes/test-wc-gateway-amazon-payments-advanced.php

```diff
@@ -0,0 +1,147 @@
+	 *
+	 * @return void
+	 */
+	public static function tear_down_after_class() : void {
+		parent::tear_down_after_class();
+		self::$gateway = null;
+		delete_option( 'woocommerce_currency' );
+		delete_option( 'woocommerce_amazon_payments_advanced_settings' );
+		delete_option( 'woocommerce_amazon_payments_new_install' );
+		delete_option( 'amazon_api_version' );
+		delete_option( WC_Amazon_Payments_Advanced_Merchant_Onboarding_Handler::KEYS_OPTION_PRIVATE_KEY );
+	}
+
+	/**
+	 * Test that gateway is not available.
+	 *
+	 * @return void
+	 */
+	public function test_is_not_available() : void {
+		$this->assertFalse( self::$gateway->is_available() );
+	}
+
+	/**
+	 * Test that gateway is available.
+	 *
+	 * @return void
+	 */
+	public function test_is_available() : void {
+		self::update_plugin_settings();
+		$this->assertTrue( self::$gateway->is_available() );
+	}
+
+	/**
+	 * Test happy path of the express gateway.
+	 *
+	 * @return void
+	 */
+	public function test_express_process_payment() : void {
+		$checkout_session_key = apply_filters( 'woocommerce_amazon_pa_checkout_session_key', 'amazon_checkout_session_id' );
+		WC()->session->set( $checkout_session_key, 'TEST_CHECKOUT_SESSION_ID' );
+		WC()->session->save_data();
+
+		$order = WC_Helper_Order::create_order( self::$gateway->id );
+
+		$mock_gateway = new WC_Mocker_Gateway_Amazon_Payments_Advanced();
+
+		$this->assertEquals(
+			array(
+				'result'   => 'success',
+				'redirect' => 'https://amazon.unit.tests/',
+			),
+			$mock_gateway->process_payment( $order->get_id() )
+		);
+	}
+
+	/**
+	 * Test happy path of classic gateway.
+	 *
+	 * @return void
+	 */
+	public function test_classic_process_payment() : void {
+		$checkout_session_key = apply_filters( 'woocommerce_amazon_pa_checkout_session_key', 'amazon_checkout_session_id' );
+		WC()->session->set( $checkout_session_key, null );
+		WC()->session->save_data();
+
+		$order_total = 100;
+
+		$order = WC_Helper_Order::create_order( self::$gateway->id, 1, $order_total );
+
+		$mock_gateway = new WC_Mocker_Gateway_Amazon_Payments_Advanced( $order_total );
+
+		remove_action( 'woocommerce_after_checkout_validation', array( wc_apa()->get_gateway(), 'classic_validation' ), 10 );
+
+		$this->assertEquals(
+			array(
+				'result'                     => 'success',
+				'redirect'                   => '#amazon-pay-classic-id-that-should-not-exist',
+				'amazonCreateCheckoutParams' => wp_json_encode( WC_Mocker_Amazon_Payments_Advanced_API::get_create_checkout_classic_session_config( array( 'test' ) ) ),
+				'amazonEstimatedOrderAmount' => wp_json_encode(
+					array(
+						'amount'       => $order_total,
+						'currencyCode' => get_woocommerce_currency(),
+					)
+				),
+			),
+			$mock_gateway->process_payment( $order->get_id() )
+		);
+	}
+
+	/**
+	 * Update plugin settings in order to make the gateway available.
+	 *
+	 * @return void
+	 */
+	protected function update_plugin_settings() : void {
```

Comment:
This method is used only in one test case, so I'd inline it and add comments marking the test setup and the test itself. The logic is not that complex, so inlining will not harm.

Reply:
agreed, done!